src/test/unit/model/model_base_crtp_test.cpp

```diff
@@ -0,0 +1,147 @@
+      else if (propto && !jacobian) return 6;
+      else return 8;
+    }
+  }
+
+  void transform_inits(const stan::io::var_context& context,
+                       Eigen::VectorXd& params_r,
+                       std::ostream* msgs) const override {
+  }
+
+  template <typename RNG>
+  void write_array(RNG& base_rng,
+                   Eigen::VectorXd& params_r,
+                   Eigen::VectorXd& params_constrained_r,
+                   bool include_tparams,
+                   bool include_gqs,
+                   std::ostream* msgs) const {
+  }
+
+  template <bool propto, bool jacobian, typename T>
+  T log_prob(std::vector<T>& params_r,
+             std::vector<int>& params_i,
+             std::ostream* msgs) const {
+    if (std::is_same<T, double>::value) {
+      if (!propto && !jacobian) return 1;
+      else if (!propto && jacobian) return 3;
+      else if (propto && !jacobian) return 5;
+      else return 7;
+    } else {
+      if (!propto && !jacobian) return 2;
+      else if (!propto && jacobian) return 4;
+      else if (propto && !jacobian) return 6;
+      else return 8;
+    }
+  }
+
+  void transform_inits(const stan::io::var_context& context,
+                       std::vector<int>& params_i,
+                       std::vector<double>& params_r,
+                       std::ostream* msgs) const override {
+  }
+
+  template <typename RNG>
+  void write_array(RNG& base_rng,
+                   std::vector<double>& params_r,
+                   std::vector<int>& params_i,
+                   std::vector<double>& params_r_constrained,
+                   bool include_tparams,
+                   bool include_gqs,
+                   std::ostream* msgs) const {
+  }
+};
+
+TEST(model, modelBaseInheritance) {
+  // check that base_model and prob_grad inheritance works
+  mock_model m(17);
+  EXPECT_EQ(17u, m.num_params_r());
+  EXPECT_EQ(0u, m.num_params_i());
+  EXPECT_THROW(m.param_range_i(0), std::out_of_range);
+}
+
+TEST(model, modelTemplateLogProb) {
+  mock_model m(17);
+  stan::model::model_base& bm = m;
+  Eigen::VectorXd params_r(2);
+  Eigen::Matrix<stan::math::var, -1, 1> params_r_v(3);
+  std::stringstream ss;
+  std::ostream* msgs = &ss;
+
+  // test from base class reference
+  EXPECT_FLOAT_EQ(1, bm.log_prob(params_r, msgs));
+  EXPECT_FLOAT_EQ(2, bm.log_prob(params_r_v, msgs).val());
+  EXPECT_FLOAT_EQ(3, bm.log_prob_jacobian(params_r, msgs));
+  EXPECT_FLOAT_EQ(4, bm.log_prob_jacobian(params_r_v, msgs).val());
+  EXPECT_FLOAT_EQ(5, bm.log_prob_propto(params_r, msgs));
+  EXPECT_FLOAT_EQ(6, bm.log_prob_propto(params_r_v, msgs).val());
+  EXPECT_FLOAT_EQ(7, bm.log_prob_propto_jacobian(params_r, msgs));
+  EXPECT_FLOAT_EQ(8, bm.log_prob_propto_jacobian(params_r_v, msgs).val());
+
+  // test template version from base class reference
+  // long form assignment avoids test macro parse error with multi tparams
+  double v1 =  bm.template log_prob<false, false>(params_r, msgs);
+  EXPECT_FLOAT_EQ(1, v1);
```

Comment:
You could use extra brackets to avoid the assignment like this:

`EXPECT_FLOAT_EQ(1, (bm.template log_prob<false, false>(params_r, msgs)));`

Same applies to the model_base_test.cpp